### PR TITLE
Improve reshapeability of models with eltwise nodes influencing shapes

### PR DESCRIPTION
**Pull request by rkazants**
Description: Reshape node always needs to be inserted in order to preserve ShapeOf nodes (reshapability of a model) that can potentially be above elementwise node.

- Refactor EltwiseInputReshape_test and EltwiseInputNormalization_test since the logic of maintaining reshape for eltwise has been changed.
- Merge EltwiseInputReshape and EltwiseInputNormalization transformations and corresponding tests
- Refactor EltwiseInputReshape transformation to move to new Graph API.
- Remove Unsqueeze from fused_op in nGraph

JIRA: 39665

Code:
* [x]  Comments
* [x]  Code style (PEP8)
* [x]  Transformation generates reshape-able IR
* [x]  Transformation preserves original framework node names

Validation:
* [x]  Unit tests
* [x]  Framework operation tests, N/A
* [x]  Transformation tests
* [x]  e2e model test with an update of ./tests/e2e_oss/utils/reshape_utils.py
* [x]  Model Optimizer IR Reader check

Documentation:
* [x]  Supported frameworks operations list, N/A
* [x]  Supported **public** models list, N/A
* [x]  New operations specification, N/A
* [x]  Guide on how to convert the **public** model, N/A
* [x]  User guide update, N/A

Signed-off-by: Roman Kazantsev <roman.kazantsev@intel.com>

### CODE REVIEW

File: model-optimizer/extensions/middle/EltwiseInputReshape.py [hunk 1]
```diff
@@ -88,35 +88,20 @@ def find_and_replace_pattern(self, graph: Graph):
...
-                        del edge_attrs['new_shape']
-
-                        # Reconnect edge from original data node to Reshape output datanode
-                        graph.remove_edge(node.id, consumer.id)
-                        graph.add_edge(reshape_data.id, consumer.id, **edge_attrs)
+            # Insert Reshape layer between data node and consumer
+            for shape_key in mapping.keys():
+                shape = list(shape_key)
+                reshape_name = node.soft_get('name', node.id) + '/EltwiseReshape'
+                reshape = Reshape(graph, attrs={'name': reshape_name})
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/EltwiseInputReshape.py:95]: use `create_op_with_const_inputs` function

File: model-optimizer/extensions/middle/EltwiseInputReshape.py [hunk 2]
```diff
@@ -88,35 +88,20 @@ def find_and_replace_pattern(self, graph: Graph):
...
+                                    {'value': shape, 'name': reshape_name + '/Shape'}).create_node_with_data()
+                reshape_data = reshape.create_node_with_data(inputs=[node, reshape_dim])
+
+                # Iterate over consumers and reconnect them to Reshape layer output
+                for consumer in mapping[shape_key]:
+                    edge_attrs = graph.get_edge_data(node.id, consumer.id)[0]
+                    del edge_attrs['new_shape']
+
+                    # Reconnect edge from original data node to Reshape output datanode
+                    graph.remove_edge(node.id, consumer.id)
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/EltwiseInputReshape.py:106]: nit: Can we rewrite this transformation using Graph API instead of low-level API?

File: model-optimizer/extensions/middle/EltwiseInputReshape.py (hunk 1 above)
Comment (rkazants) [reply model-optimizer/extensions/middle/EltwiseInputReshape.py:95]: this is not suitable. It needs simultaneous creation of data node for `Reshape`.

File: model-optimizer/extensions/middle/EltwiseInputReshape.py (hunk 2 above)
Comment (rkazants) [reply model-optimizer/extensions/middle/EltwiseInputReshape.py:106]: Yes, it is preferable but let us leave it for another time.

File: model-optimizer/extensions/middle/EltwiseInputReshape.py [hunk 3]
```diff
@@ -88,7 +88,7 @@ def find_and_replace_pattern(self, graph: Graph):
                     else:
                         mapping[new_shape].append(consumer)
 
-            if node.has_valid('value'):
+            if node.has_valid('value') and node.has_valid("type") and node.type == "Const":
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/EltwiseInputReshape.py:91]: I still would like to see only the solution with adding "Reshape" operation + update the code with a Graph API instead of low-level graph modification.

File: model-optimizer/mo/utils/shape.py [hunk 4]
```diff
@@ -38,13 +38,13 @@ def get_canonical_axis_index_node(rank: Node, axis: int) -> Node:
     graph = rank.graph
     name = rank.soft_get('name', rank.id)
     if axis < 0:
-        axis = Const(graph, {'name': name + '/negative_axis', 'value': int64_array([axis])}).create_node()
```
Comment (lazarevevgeny) [inline model-optimizer/mo/utils/shape.py:41]: The function `get_canonical_axis_index_node` is used in the `AtenToEmbeddingBag` and `MXRepeatReplacer` transformations. The fact that you changed it here may result in the different behaviour of these transformation.

Comment (rkazants) [reply model-optimizer/mo/utils/shape.py:41]: Yes, I know but it is wrong for them. As I understand e2e tests with them pass. Now we have a new issue with `Unsquueze`.
To implement a new transformation for `Range` with 1d input tensors seems wrong to me since that `Range` is created by MO and this is not MXNet specifics

File: model-optimizer/extensions/middle/EltwiseInputReshape.py (hunk 3 above)
Comment (rkazants) [reply model-optimizer/extensions/middle/EltwiseInputReshape.py]: Done.

File: model-optimizer/extensions/middle/EltwiseInputReshape.py [hunk 5]
```diff
@@ -73,50 +74,38 @@ def run_after(self):
...
-                edge_attrs = graph.get_edge_data(node.id, consumer.id)[0]
-                if 'new_shape' in edge_attrs:
-                    if np.array_equal(edge_attrs['new_shape'], node.shape):
-                        continue
-                    new_shape = tuple([x for x in edge_attrs['new_shape']])
-                    if not new_shape in mapping:
-                        mapping.update({new_shape: [consumer]})
-                    else:
-                        mapping[new_shape].append(consumer)
+        for node in graph.get_op_nodes():
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/EltwiseInputReshape.py:77]: Marking a node edge with 'new_shape' attribute in one transformation then calling directly this one to insert Reshape looks awkward. Also even a new code looks ugly. So I have a revolutionary suggestion. Let's merge `EltwiseInputReshape` and `EltwiseInputNormalize` into one transformation (keep `EltwiseInputReshape`) and let's insert `Reshape` operations "on the fly" when we detect that it is needed. It will make code much smaller + we will avoid redundant graph traversals and checking of all graph edges attributes (which is also time consuming).

**Review by lazarevevgeny — CHANGES_REQUESTED** [review]

File: model-optimizer/extensions/middle/EltwiseInputReshape.py (hunk 5 above)
Comment (rkazants) [reply model-optimizer/extensions/middle/EltwiseInputReshape.py:72]: Done.

File: model-optimizer/extensions/middle/EltwiseInputReshape.py [hunk 6]
```diff
@@ -65,58 +65,46 @@ def find_and_replace_pattern(self, graph: Graph):
...
+            for out_port_idx in node.out_ports():
+                mapping = {}
+                output_port = node.out_port(out_port_idx)
+                output_port_shape = output_port.data.get_shape()
+                for consumer_port in output_port.get_destinations():
+                    if consumer_port.node.has_valid('is_eltwise') and consumer_port.node['is_eltwise'] == True:
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/EltwiseInputReshape.py:78]: Replace with `if consumer_port.node.has_and_set('is_eltwise'):`

File: model-optimizer/extensions/middle/EltwiseInputReshape.py [hunk 7]
```diff
@@ -65,58 +65,46 @@ def find_and_replace_pattern(self, graph: Graph):
...
+                        consumer_output_shape = consumer_port.node.out_port(0).data.get_shape()
+                        if np.array_equal(consumer_output_shape, output_port_shape):
+                            continue
+                        # Set edge attribute new_shape for further transformation pass
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/EltwiseInputReshape.py:82]: Comment is outdated

File: model-optimizer/extensions/middle/EltwiseInputReshape.py [hunk 8]
```diff
@@ -65,58 +65,46 @@ def find_and_replace_pattern(self, graph: Graph):
...
+                        new_shape = output_port_shape
+                        if len(output_port_shape) != len(consumer_output_shape):
+                            for x in range(len(consumer_output_shape) - len(output_port_shape)):
+                                new_shape = np.insert(new_shape, 0, 1)
+                        new_shape = tuple([x for x in new_shape])
+                        if not new_shape in mapping:
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/EltwiseInputReshape.py:88]: Why do you save them in the `mapping`? You can directly insert the Reshape operation when you detected the eltwise which input should be reshaped.

**Review by lazarevevgeny — CHANGES_REQUESTED** [review]

File: model-optimizer/extensions/middle/EltwiseInputReshape.py (hunk 8 above)
Comment (rkazants) [reply model-optimizer/extensions/middle/EltwiseInputReshape.py:88]: The mapping is created to know for which eltwise consumers the same normalization (i.e. Reshape with output shape) is required. According to your way we can have Reshapes with duplicate output shapes that is not optimal.

File: model-optimizer/extensions/middle/EltwiseInputReshape.py (hunk 6 above)
Comment (rkazants) [reply model-optimizer/extensions/middle/EltwiseInputReshape.py]: Fixed.

File: model-optimizer/extensions/middle/EltwiseInputReshape.py (hunk 7 above)
Comment (rkazants) [reply model-optimizer/extensions/middle/EltwiseInputReshape.py]: Fixed.

File: model-optimizer/extensions/middle/EltwiseInputReshape.py [hunk 9]
```diff
@@ -65,58 +65,48 @@ def find_and_replace_pattern(self, graph: Graph):
...
-                edge_attrs = graph.get_edge_data(node.id, consumer.id)[0]
-                if 'new_shape' in edge_attrs:
-                    if np.array_equal(edge_attrs['new_shape'], node.shape):
-                        continue
-                    new_shape = tuple([x for x in edge_attrs['new_shape']])
-                    if not new_shape in mapping:
-                        mapping.update({new_shape: [consumer]})
-                    else:
-                        mapping[new_shape].append(consumer)
+        for node in graph.get_op_nodes():
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/EltwiseInputReshape.py:72]: It is much more efficient to iterate over all `is_eltwise` operations here and check their input ports shapes and output port shape. Currently you iterate over all graph nodes, then for each output port of the operation you get the consumers and if the consumer is eltwise operation do the magic. Very inefficient.

File: model-optimizer/extensions/middle/EltwiseInputReshape.py [hunk 10]
```diff
@@ -65,58 +65,48 @@ def find_and_replace_pattern(self, graph: Graph):
...
                 # Insert Reshape layer between data node and consumer
                 for shape_key in mapping.keys():
-                    shape = list(shape_key)
+                    new_shape = list(shape_key)
                     reshape_name = node.soft_get('name', node.id) + '/EltwiseReshape'
-                    reshape = Reshape(graph, attrs={'name': reshape_name})
-                    reshape_dim = Const(graph,
-                                        {'value': shape, 'name': reshape_name + '/Shape'}).create_node_with_data()
-                    reshape_data = reshape.create_node_with_data(inputs=[node, reshape_dim])
+                    reshape_node = create_op_with_const_inputs(graph, Reshape, {1: new_shape},
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/EltwiseInputReshape.py:99]: The Reshape must be replaced with "Unsqueeze". By inserting Reshape with a hardcoded shape we loose reshape-ability.

File: model-optimizer/extensions/middle/EltwiseInputReshape.py [hunk 11]
```diff
@@ -65,58 +65,48 @@ def find_and_replace_pattern(self, graph: Graph):
...
+                mapping = {}
+                output_port = node.out_port(out_port_idx)
+                output_port_shape = output_port.data.get_shape()
+                for consumer_port in output_port.get_destinations():
+                    if consumer_port.node.has_and_set('is_eltwise'):
+                        consumer_output_shape = consumer_port.node.out_port(0).data.get_shape()
+                        if np.array_equal(consumer_output_shape, output_port_shape):
+                            continue
+                        # Compute a new shape that is required for a shape normalization
+                        new_shape = output_port_shape
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/EltwiseInputReshape.py:85]: `new_shape = output_port_shape` -> `new_shape = output_port_shape.copy()`

**Review by lazarevevgeny — CHANGES_REQUESTED** [review]

File: model-optimizer/extensions/middle/EltwiseInputReshape.py (hunk 10 above)
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/EltwiseInputReshape.py:99]: Also since you modify the output shape of the operation you probably need the "override_output_shape" attribute and set `force_shape_inference = True` in the transformation

File: model-optimizer/extensions/middle/EltwiseInputReshape.py (hunk 9 above)
Comment (rkazants) [reply model-optimizer/extensions/middle/EltwiseInputReshape.py]: Fixed. As we agreed now it is two pass transformation: generate a map for eltwise producers with non-normalized shapes; walkthrough the map and insert reshape.

File: model-optimizer/extensions/middle/EltwiseInputReshape.py (hunk 10 above)
Comment (rkazants) [reply model-optimizer/extensions/middle/EltwiseInputReshape.py]: Great catch. I will fix it.

File: model-optimizer/extensions/middle/EltwiseInputReshape.py (hunk 11 above)
Comment (rkazants) [reply model-optimizer/extensions/middle/EltwiseInputReshape.py]: Fixed.

File: model-optimizer/extensions/middle/EltwiseInputReshape.py (hunk 10 above)
Comment (rkazants) [reply model-optimizer/extensions/middle/EltwiseInputReshape.py]: Fixed.

Comment (rkazants) [reply model-optimizer/extensions/middle/EltwiseInputReshape.py]: Done.

File: model-optimizer/extensions/middle/EltwiseInputReshape.py [hunk 12]
```diff
@@ -65,58 +65,50 @@ def find_and_replace_pattern(self, graph: Graph):
...
+                consumer_port = eltwise_node.in_port(in_port_idx)
+                producer_port = consumer_port.get_source()
+                producer_shape = producer_port.data.get_shape()
+                if len(producer_shape) == len(eltwise_shape):
+                    continue
+                unsqueeze_dims = []
+                if len(producer_shape) != len(eltwise_shape):
+                    for x in range(len(eltwise_shape) - len(producer_shape)):
+                        unsqueeze_dims.append(x)
+                unsqueeze_dims = tuple([x for x in unsqueeze_dims])
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/EltwiseInputReshape.py:89]: unsqueeze_dims = tuple(np.arange(len(eltwise_shape) - len(producer_shape), dtype=np.int64))

File: model-optimizer/extensions/middle/EltwiseInputReshape.py [hunk 13]
```diff
@@ -65,58 +65,50 @@ def find_and_replace_pattern(self, graph: Graph):
...
+                if not producer_port in mapping:
+                    mapping.update({producer_port: {unsqueeze_dims: [consumer_port]}})
+                else:
+                    int_mapping = mapping[producer_port].copy()
+                    if not unsqueeze_dims in int_mapping:
+                        int_mapping.update({unsqueeze_dims: [consumer_port]})
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/EltwiseInputReshape.py:95]: Move this code undef `if` statement when `len(producer_shape) != len(eltwise_shape)`

File: model-optimizer/extensions/middle/EltwiseInputReshape.py [hunk 14]
```diff
@@ -65,58 +65,50 @@ def find_and_replace_pattern(self, graph: Graph):
...
+                    continue
+                unsqueeze_dims = []
+                if len(producer_shape) != len(eltwise_shape):
+                    for x in range(len(eltwise_shape) - len(producer_shape)):
+                        unsqueeze_dims.append(x)
+                unsqueeze_dims = tuple([x for x in unsqueeze_dims])
+                if not producer_port in mapping:
+                    mapping.update({producer_port: {unsqueeze_dims: [consumer_port]}})
+                else:
+                    int_mapping = mapping[producer_port].copy()
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/EltwiseInputReshape.py:93]: Why do you create a copy here? I think that we can directly update the `int_mapping` dictionary.

File: model-optimizer/extensions/middle/EltwiseInputReshape.py [hunk 15]
```diff
@@ -65,58 +65,50 @@ def find_and_replace_pattern(self, graph: Graph):
...
-
-                        # Reconnect edge from original data node to Reshape output datanode
-                        graph.remove_edge(node.id, consumer.id)
-                        graph.add_edge(reshape_data.id, consumer.id, **edge_attrs)
+                        int_mapping[unsqueeze_dims].append(consumer_port)
+                    mapping[producer_port] = int_mapping
+
+        # Walkthough each produced in the map and insert Reshape nodes between a producer and eltwise nodes
+        for producer_port in mapping.keys():
+            int_mapping = mapping[producer_port].copy()
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/EltwiseInputReshape.py:102]: Why copy? We can iterate over the elements without creating a copy.

File: model-optimizer/extensions/middle/EltwiseInputReshape.py [hunk 16]
```diff
@@ -65,58 +65,50 @@ def find_and_replace_pattern(self, graph: Graph):
...
+            producer_node = producer_port.node
+            for unsqueeze_dims in int_mapping.keys():
+                reshape_name = producer_node.soft_get('name', producer_node.id) + '/EltwiseReshape'
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/EltwiseInputReshape.py:105]: nit: may be rename to `unsqueeze_name`?

**Review by lazarevevgeny — CHANGES_REQUESTED** [review]

File: ngraph/core/include/ngraph/op/unsqueeze.hpp [hunk 17]
```diff
@@ -55,5 +50,3 @@ namespace ngraph
         using v0::Unsqueeze;
```
Comment (lazarevevgeny) [inline ngraph/core/include/ngraph/op/unsqueeze.hpp:50]: please, remove this line and update places where `op::Unsqueeze` is used instead of `op::v0::Unsqueeze`.

@ilyachur , am I right?

**Review by lazarevevgeny — CHANGES_REQUESTED** [review]

File: ngraph/core/include/ngraph/op/unsqueeze.hpp (hunk 17 above)
Comment (ilyachur) [reply ngraph/core/include/ngraph/op/unsqueeze.hpp:50]: Yes you are right!

File: model-optimizer/extensions/middle/EltwiseInputReshape.py (hunk 12 above)
Comment (rkazants) [reply model-optimizer/extensions/middle/EltwiseInputReshape.py]: Done.

File: model-optimizer/extensions/middle/EltwiseInputReshape.py (hunk 13 above)
Comment (rkazants) [reply model-optimizer/extensions/middle/EltwiseInputReshape.py]: Done.

File: model-optimizer/extensions/middle/EltwiseInputReshape.py (hunk 14 above)
Comment (rkazants) [reply model-optimizer/extensions/middle/EltwiseInputReshape.py]: Done.

File: model-optimizer/extensions/middle/EltwiseInputReshape.py (hunk 15 above)
Comment (rkazants) [reply model-optimizer/extensions/middle/EltwiseInputReshape.py]: Done.

File: model-optimizer/extensions/middle/EltwiseInputReshape.py (hunk 16 above)
Comment (rkazants) [reply model-optimizer/extensions/middle/EltwiseInputReshape.py]: Done.

File: ngraph/core/include/ngraph/op/unsqueeze.hpp (hunk 17 above)
Comment (rkazants) [reply ngraph/core/include/ngraph/op/unsqueeze.hpp]: Done.

File: model-optimizer/extensions/middle/EltwiseInputReshape.py [hunk 18]
```diff
@@ -65,58 +67,41 @@ def find_and_replace_pattern(self, graph: Graph):
...
 class EltwiseInputReshape(MiddleReplacementPattern):
-    enabled = True
-    force_clean_up = True
-
-    def run_after(self):
-        from extensions.middle.pass_separator import MiddleStart
-        return [MiddleStart]
+    # This pass should be called directly from pipeline before layout change and other permutations
+    enabled = False
+    force_shape_inference = True
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/EltwiseInputReshape.py:72]: `force_shape_inference` does not make sense because the transformation is disabled.

Comment (lazarevevgeny) [reply model-optimizer/extensions/middle/EltwiseInputReshape.py:72]: Looks like we need to restore the original solution with calling `infer` method for the ops which will change the output shape.

**Review by lazarevevgeny — CHANGES_REQUESTED** [review]

File: model-optimizer/extensions/middle/EltwiseInputReshape.py (hunk 18 above)
Comment (rkazants) [reply model-optimizer/extensions/middle/EltwiseInputReshape.py]: Fixed.

**Review by lazarevevgeny — APPROVED** [review]
MO changes looks good

**Comment by rkazants** [comment]
Now I anticipate to get https://github.com/opencv/opencv/pull/18741 merged to make pre-commit green.

**Review by ilyachur — APPROVED** [review]
nGraph part LGTM

File: ngraph/core/include/ngraph/op/unsqueeze.hpp [hunk 19]
```diff
@@ -52,8 +47,5 @@ namespace ngraph
                     clone_with_new_inputs(const OutputVector& new_args) const override;
             };
         }
-        using v0::Unsqueeze;
```
Comment (ilyachur) [inline ngraph/core/include/ngraph/op/unsqueeze.hpp:55]: I have thought more about this using. And it is not correct that we remove it.

If we remove it we break the backward compatibility. It is why I recommend to revert this change. Sorry for inconvinience.

We can remove `using` only for deprecated operations.

Comment (rkazants) [reply ngraph/core/include/ngraph/op/unsqueeze.hpp:55]: Fixed.

File: model-optimizer/extensions/middle/EltwiseInputReshape.py [hunk 20]
```diff
@@ -65,6 +67,69 @@ def find_and_replace_pattern(self, graph: Graph):
 
 
 class EltwiseInputReshape(MiddleReplacementPattern):
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/EltwiseInputReshape.py:69]: We can change this to just a function. No need to use a class here.

File: model-optimizer/extensions/middle/EltwiseInputReshape.py [hunk 21]
```diff
@@ -65,6 +67,69 @@ def find_and_replace_pattern(self, graph: Graph):
...
+                else:
+                    continue
+                if not producer_port in mapping:
+                    mapping.update({producer_port: {unsqueeze_dims: [consumer_port]}})
+                elif not unsqueeze_dims in mapping[producer_port]:
+                    mapping[producer_port].update({unsqueeze_dims: [consumer_port]})
+                else:
+                    mapping[producer_port][unsqueeze_dims].append(consumer_port)
+
+        # Walk through each produced in the map and insert Reshape nodes between a producer and eltwise nodes
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/EltwiseInputReshape.py:101]: nit: Reshape -> Unsqueeze

File: model-optimizer/extensions/middle/EltwiseInputReshape.py [hunk 22]
```diff
@@ -65,6 +67,69 @@ def find_and_replace_pattern(self, graph: Graph):
...
+                # Insert Reshape with determined output shape between the current producer and eltwise node
+                for consumer_port in mapping[producer_port][unsqueeze_dims]:
+                    consumer_port.connect(unsqueeze_node.out_port(0))
+
+                # The shape and value adjustments must be explicitly done within the transformation
+                # since the transformation is called from Fusing transformation that excludes
+                # automatic call of shape inference pass
+                producer_port_value = producer_port.data.get_value()
+                producer_port_shape = producer_port.data.get_shape()
+                new_shape = producer_port_shape
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/EltwiseInputReshape.py:120]: `new_shape = producer_port_shape.copy()`

File: model-optimizer/extensions/middle/EltwiseInputReshape.py [hunk 23]
```diff
@@ -65,6 +67,69 @@ def find_and_replace_pattern(self, graph: Graph):
...
+                elif not unsqueeze_dims in mapping[producer_port]:
+                    mapping[producer_port].update({unsqueeze_dims: [consumer_port]})
+                else:
+                    mapping[producer_port][unsqueeze_dims].append(consumer_port)
+
+        # Walk through each produced in the map and insert Reshape nodes between a producer and eltwise nodes
+        for producer_port in mapping.keys():
+            producer_node = producer_port.node
+            for unsqueeze_dims in mapping[producer_port].keys():
+                unsqueeze_name = producer_node.soft_get('name', producer_node.id) + '/EltwiseReshape'
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/EltwiseInputReshape.py:105]: nit: EltwiseReshape -> EltwiseUnsqueeze

File: model-optimizer/extensions/middle/EltwiseInputReshape.py [hunk 24]
```diff
@@ -65,6 +67,69 @@ def find_and_replace_pattern(self, graph: Graph):
...
+                unsqueeze_node = create_op_with_const_inputs(graph, Unsqueeze, {1: int64_array(list(unsqueeze_dims))},
+                                                             {'name': unsqueeze_name})
+
+                unsqueeze_node.in_port(0).connect(producer_port)
+
+                # Insert Reshape with determined output shape between the current producer and eltwise node
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/EltwiseInputReshape.py:111]: Unsqueeze

File: model-optimizer/extensions/middle/EltwiseInputReshape.py [hunk 25]
```diff
@@ -65,6 +67,69 @@ def find_and_replace_pattern(self, graph: Graph):
...
+
+                producer_data_node = eltwise_node.in_node(in_port_idx)
+                edge_attrs = graph.get_edge_data(producer_data_node.id, eltwise_node.id)[0]
+                if is_first_pass and 'unsqueeze_dims' in edge_attrs and len(edge_attrs['unsqueeze_dims']) > 0:
+                    unsqueeze_dims = tuple([x for x in edge_attrs['unsqueeze_dims']])
+                elif not is_first_pass and len(producer_shape) != len(eltwise_shape):
+                    unsqueeze_dims = tuple(np.arange(len(eltwise_shape) - len(producer_shape), dtype=np.int64))
+                else:
+                    continue
+                if not producer_port in mapping:
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/EltwiseInputReshape.py:94]: `if producer_port not in mapping:`

File: model-optimizer/extensions/middle/EltwiseInputReshape.py [hunk 26]
```diff
@@ -65,6 +67,69 @@ def find_and_replace_pattern(self, graph: Graph):
...
+                    mapping.update({producer_port: {unsqueeze_dims: [consumer_port]}})
+                elif not unsqueeze_dims in mapping[producer_port]:
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/EltwiseInputReshape.py:96]: `elif unsqueeze_dims not in mapping[producer_port]:`

File: model-optimizer/extensions/middle/EltwiseInputReshape.py (hunk 20 above)
Comment (lazarevevgeny) [reply model-optimizer/extensions/middle/EltwiseInputReshape.py:69]: Also, please explain in the function comment what `is_first_pass` mean.

File: model-optimizer/extensions/middle/EltwiseInputReshape.py [hunk 27]
```diff
@@ -65,6 +67,69 @@ def find_and_replace_pattern(self, graph: Graph):
...
+                new_shape = producer_port_shape
+                for unsqueeze_dim in unsqueeze_dims:
+                    new_shape = np.insert(new_shape, unsqueeze_dim, 1)
+                if producer_port_value is not None:
+                    unsqueeze_node.out_port(0).data.set_value(np.reshape(producer_port_value, new_shape))
+                else:
+                    unsqueeze_node.out_port(0).data.set_shape(new_shape)
+
+
+class EltwiseInputReshapeFirstPass(MiddleReplacementPattern):
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/EltwiseInputReshape.py:129]: Can we rename the transformation to be more meaningful? "FirstPass" is not intuitive...

**Review by lazarevevgeny — CHANGES_REQUESTED** [review]

File: model-optimizer/extensions/middle/EltwiseInputReshape.py [hunk 28]
```diff
@@ -64,59 +63,86 @@ def find_and_replace_pattern(self, graph: Graph):
...
+# compute a map of unsqueeze_dims for each producer of eltwise node
+def compute_unsqueeze_map_for_eltwise(eltwise_node: Node):
+    eltwise_shape = eltwise_node.out_port(0).data.get_shape()
+    max_dims = None
+    for in_port_idx in eltwise_node.in_ports():
+        consumer_port = eltwise_node.in_port(in_port_idx)
+        producer_port = consumer_port.get_source()
+        producer_shape = producer_port.data.get_shape()
+        if max_dims is None or len(producer_shape) > max_dims:
+            max_dims = len(producer_shape)
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/EltwiseInputReshape.py:75]: max_dims = max([port.data.get_shape() for port in eltwise_node.in_ports().values() if port.data.get_shape() is not None])

File: model-optimizer/extensions/middle/EltwiseInputReshape.py [hunk 29]
```diff
@@ -64,59 +63,86 @@ def find_and_replace_pattern(self, graph: Graph):
...
+
+    axis = eltwise_node.axis if eltwise_node.has_valid('axis') else None
+    unsqueeze_dims_map = {}
+    for in_port_idx in eltwise_node.in_ports():
+        consumer_port = eltwise_node.in_port(in_port_idx)
+        producer_port = consumer_port.get_source()
+        producer_shape = producer_port.data.get_shape()
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/EltwiseInputReshape.py:82]:
```
for in_port_idx, consumer_port in eltwise_node.in_ports().items():
    producer_port = consumer_port.get_source()
    producer_shape = consumer_port.data.get_shape()
```

File: model-optimizer/extensions/middle/EltwiseInputReshape.py [hunk 30]
```diff
@@ -64,59 +63,86 @@ def find_and_replace_pattern(self, graph: Graph):
...
+    eltwise_shape = eltwise_node.out_port(0).data.get_shape()
+    max_dims = None
+    for in_port_idx in eltwise_node.in_ports():
+        consumer_port = eltwise_node.in_port(in_port_idx)
+        producer_port = consumer_port.get_source()
+        producer_shape = producer_port.data.get_shape()
+        if max_dims is None or len(producer_shape) > max_dims:
+            max_dims = len(producer_shape)
+
+    axis = eltwise_node.axis if eltwise_node.has_valid('axis') else None
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/EltwiseInputReshape.py:77]: `axis = eltwise_node.soft_get('axis', None)`

File: model-optimizer/extensions/middle/EltwiseInputReshape.py [hunk 31]
```diff
@@ -64,59 +63,86 @@ def find_and_replace_pattern(self, graph: Graph):
...
-                    for consumer in mapping[shape_key]:
-                        edge_attrs = graph.get_edge_data(node.id, consumer.id)[0]
-                        del edge_attrs['new_shape']
-
-                        # Remove edge from previous data node and connect new data node with its consumer
-                        graph.remove_edge(node.id, consumer.id)
-                        graph.add_edge(node_copy.id, consumer.id, **edge_attrs)
+# compute a map of unsqueeze_dims for each producer of eltwise node
+def compute_unsqueeze_map_for_eltwise(eltwise_node: Node):
+    eltwise_shape = eltwise_node.out_port(0).data.get_shape()
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/EltwiseInputReshape.py:68]: nit: rename to `output_shape`

File: model-optimizer/extensions/middle/EltwiseInputReshape.py [hunk 32]
```diff
@@ -64,59 +63,86 @@ def find_and_replace_pattern(self, graph: Graph):
...
+# normalize input shapes for eltwise node by using Unsqueeze operation
+def normalize_eltwise_inputs(graph: Graph):
+    # Generate a map for producers of eltwise nodes with non-normalized shapes
+    # and in this map every producer has another map that reflects normalized shape
+    # to a list of eltwise consumers
+    mapping = {}
+    for eltwise_node in graph.get_op_nodes(is_eltwise=True):
+        unsqueeze_dims_map = compute_unsqueeze_map_for_eltwise(eltwise_node)
+        for in_port_idx in eltwise_node.in_ports():
+            consumer_port = eltwise_node.in_port(in_port_idx)
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/EltwiseInputReshape.py:112]: `for in_port_idx, consumer_port in eltwise_node.in_ports().items():`

**Review by lazarevevgeny — COMMENTED** [review]
In general, code looks good, but some improvements should be applied

File: model-optimizer/extensions/middle/EltwiseInputReshape.py [hunk 33]
```diff
@@ -64,59 +63,86 @@ def find_and_replace_pattern(self, graph: Graph):
...
-                    new_value = np.reshape(node.value, shape)
-                    node_copy = Op.create_input_data_node(graph, node.id + '/copy', value=np.array(new_value))
-                    for consumer in mapping[shape_key]:
-                        edge_attrs = graph.get_edge_data(node.id, consumer.id)[0]
-                        del edge_attrs['new_shape']
-
-                        # Remove edge from previous data node and connect new data node with its consumer
-                        graph.remove_edge(node.id, consumer.id)
-                        graph.add_edge(node_copy.id, consumer.id, **edge_attrs)
+# compute a map of unsqueeze_dims for each producer of eltwise node
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/EltwiseInputReshape.py:66]: Please, make a comment according to our standard approach using """.

File: model-optimizer/extensions/middle/EltwiseInputReshape.py (hunk 31 above)
Comment (rkazants) [reply model-optimizer/extensions/middle/EltwiseInputReshape.py:68]: `output_shape` does not say about node of which this output is. I like `eltwise_shape` more.

**Comment by rkazants** [comment]
https://dlbenchmark-ci-icv.inn.intel.com/job/DL-Benchmark/job/try/job/ubuntu18_i7_8700/1763/

File: model-optimizer/extensions/middle/EltwiseInputReshape.py (hunk 33 above)
Comment (rkazants) [reply model-optimizer/extensions/middle/EltwiseInputReshape.py]: Done.

File: model-optimizer/extensions/middle/EltwiseInputReshape.py (hunk 32 above)
Comment (rkazants) [reply model-optimizer/extensions/middle/EltwiseInputReshape.py]: Done.

File: model-optimizer/extensions/middle/EltwiseInputReshape.py (hunk 30 above)
Comment (rkazants) [reply model-optimizer/extensions/middle/EltwiseInputReshape.py]: Done.

File: model-optimizer/extensions/middle/EltwiseInputReshape.py (hunk 29 above)
Comment (rkazants) [reply model-optimizer/extensions/middle/EltwiseInputReshape.py]: Done.

File: model-optimizer/extensions/middle/EltwiseInputReshape.py (hunk 28 above)
Comment (rkazants) [reply model-optimizer/extensions/middle/EltwiseInputReshape.py]: Done.

**Review by lazarevevgeny — APPROVED** [review]

**Review by lazarevevgeny — APPROVED** [review]
